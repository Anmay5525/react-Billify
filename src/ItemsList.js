import Table from "./Table.js";
import Loader from "./Loader.js";

const { useEffect, useState } = React;

const ItemsList = ({ handleSubRouteChange }) => {
  const [items, setItems] = useState(null);

  const controller = new AbortController();
  const signal = controller.signal;

  const getItems = () => {
    fetch("https://rzp-training.herokuapp.com/team1/items", { signal })
      .then((res) => res.json())
      .then((r) => setItems(r))
      .catch((er) => console.log(er));
  };
  useEffect(() => {
    getItems();
    return () => {
      controller.abort();
    };
  }, []);
  const fields = ["name", "description", "amount", "currency"];
  if (items) {
    const data = items.items.map((item) => {
      return {
        name: item.name,
        description: item.description,
        amount: item.amount / 100,
        currency: item.currency,
      };
    });

    return (
      <div className="content">
        <div className="items-title-container">
          <div className="items-title">Items</div>
          <button
            className="items-new-btn"
            onClick={() => handleSubRouteChange("new")}
          >
            + New Item
          </button>
        </div>
        <Table fields={fields} data={data} />
      </div>
    );
  } else {
    return <Loader />;
  }
};
export default ItemsList;
