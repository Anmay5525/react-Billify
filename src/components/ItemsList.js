import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";

import Table from "./Table";
import Loader from "./Loader";

const ItemsList = () => {
  const [items, setItems] = useState(null);

  const controller = new AbortController();
  const { signal } = controller;

  const getItems = () => {
    fetch("https://rzp-training.herokuapp.com/team1/items", { signal })
      .then((res) => res.json())
      .then((r) => setItems(r))
      .catch((er) => console.log(er));
  };
  useEffect(() => {
    getItems();
    return () => {
      controller.abort();
    };
  }, []);
  const fields = ["name", "description", "amount", "currency"];
  if (items) {
    const data = items.items.map((item) => {
      return {
        name: item.name,
        description: item.description,
        amount: item.amount / 100,
        currency: item.currency,
      };
    });

    return (
      <div className="content">
        <div className="items-title-container">
          <div className="items-title">Items</div>
          <Link to="/Items/New">
            <button
              type="button"
              className="items-new-btn"
              // onClick={() => handleSubRouteChange("new")}
            >
              + New Item
            </button>
          </Link>
        </div>
        <Table fields={fields} data={data} />
      </div>
    );
  }
  return <Loader />;
};

export default ItemsList;
