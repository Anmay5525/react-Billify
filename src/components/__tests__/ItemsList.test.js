import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import ItemsList from '../ItemsList';
import { BrowserRouter as Router } from "react-router-dom";

const itemsListData = {
  "entity": "collection",
  "count": 2,
  "items": [
    {
      "id": "item_FfjVoCpn6hG1Wu",
      "active": true,
      "name": "Kora Kagaz",
      "description": "",
      "amount": 1000,
      "unit_amount": 1000,
      "currency": "INR",
      "type": "invoice",
      "unit": null,
      "tax_inclusive": false,
      "hsn_code": null,
      "sac_code": null,
      "tax_rate": null,
      "tax_id": null,
      "tax_group_id": null,
      "created_at": 1600702921
    },
    {
      "id": "item_Ffd9vOPr1yKiUC",
      "active": true,
      "name": "disc",
      "description": "disc",
      "amount": 30000,
      "unit_amount": 30000,
      "currency": "INR",
      "type": "invoice",
      "unit": null,
      "tax_inclusive": false,
      "hsn_code": null,
      "sac_code": null,
      "tax_rate": null,
      "tax_id": null,
      "tax_group_id": null,
      "created_at": 1600680549
    },
  ]
};

describe('Testing ItemsList component if it', () => {

  beforeEach(() => {
    fetch.resetMocks();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ItemsList></ItemsList>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders loader', () => {
    const tree = renderer.create(<ItemsList></ItemsList>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('makes fetch calls', async () => {
    fetch.mockResponseOnce(JSON.stringify(itemsListData));
    const container = mount(<Router><ItemsList /></Router>);
    await act(async () => {
      (res) => setTimeout(res, 0);
    });
    container.update();
    expect(container.html()).toMatchSnapshot();
    expect(fetch).toHaveBeenCalledTimes(1);
  });

  it('handles fetch errors', async () => {
    fetch.mockRejectOnce(['Error', {status: 200}]);
    const container = mount(<Router><ItemsList /></Router>);
    await act(async () => {
      (res) => setTimeout(res, 0);
    });
    container.update();
    expect(container.html()).toMatchSnapshot();
  });

});