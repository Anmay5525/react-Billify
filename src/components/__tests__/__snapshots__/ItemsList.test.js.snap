// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ItemsList component if it handles fetch errors 1`] = `"<div class=\\"loader-container\\"><svg viewBox=\\"0 0 100 100\\" xmlns=\\"http://www.w3.org/2000/svg\\"><circle cx=\\"50\\" cy=\\"50\\" r=\\"45\\"></circle></svg></div>"`;

exports[`Testing ItemsList component if it makes fetch calls 1`] = `"<div class=\\"content\\"><div class=\\"items-title-container\\"><div class=\\"items-title\\">Items</div><a href=\\"/Items/New\\"><button type=\\"button\\" class=\\"items-new-btn\\">+ New Item</button></a></div><div><table class=\\"table\\"><thead><tr><th>NAME</th><th>DESCRIPTION</th><th>AMOUNT</th><th>CURRENCY</th></tr></thead><tbody><tr class=\\"tr\\"><td>Kora Kagaz</td><td></td><td>10</td><td>INR</td></tr><tr class=\\"tr\\"><td>disc</td><td>disc</td><td>300</td><td>INR</td></tr></tbody></table></div></div>"`;

exports[`Testing ItemsList component if it renders loader 1`] = `
<div
  className="loader-container"
>
  <svg
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="50"
      cy="50"
      r="45"
    />
  </svg>
</div>
`;
